sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1032,6 +1044,42 @@ object ColumnPruning extends Rule[LogicalPlan] {
         p
       }
 
+    case p @ Project(_, ul: UnionLoop) =>
+      if (!ul.outputSet.subsetOf(p.references)) {
+        val newAnchorChildProj = prunedChild(ul.anchor, p.references)
+        val neededIndicesListRef = {
```

Comment:
can you briefly describe the logic here? I thought we could follow how Union is handled
```
val anchor = ul.anchor
val newAnchor = prunedChild(anchor, p.references)
val newOutput = newAnchor.output
val selected = ul.recursion.output.zipWithIndex.filter { case (a, i) =>
  newOutput.contains(anchor.output(i))
}.map(_._1)
val newRecursion = Project(selected, ul.recursion)
p.copy(child = ul.withNewChildren(Seq(newAnchor, newRecursion)))
```